Add toggle_background_opacity option

aw1875 opened the pull request:
Linux implementation for https://github.com/ghostty-org/ghostty/issues/5047 (mentioning it rather than considering it closed due to the fact that I don't have access to a Mac currently to implement and test that side of this card)

aw1875 committed:
apprt/gtk: add toggle_background_opacity option

AlexJuca commented:
@aw1875 If you feel comfortable I could implement the macOS functionality that builds on your changes.

aw1875 commented:
@AlexJuca, yeah that'd be great if you wanted to do that. I'm pretty sure I know how it should look so you can definitely work on that!

AlexJuca commented:
> @AlexJuca, yeah that'd be great if you wanted to do that. I'm pretty sure I know how it should look so you can definitely work on that!

Awesome, I'll work on this tomorrow.

pluiedev commented on src/apprt/gtk/Window.zig:
```diff
@@ -559,6 +559,17 @@ pub fn toggleWindowDecorations(self: *Window) void {
+        c.gtk_widget_add_css_class(@ptrCast(self.window), "background");
+    }
+}
+
```
This doesn't actually update the config for the OpenGL renderer, so the actual terminal screen will still render as if they're transparent (and likely have wrong colors)

aw1875 replied:
Good point, didn't even think about that. Let me take a look at that as well

jcollie commented:
@aw1875 is this something that you plan on updating? There has been a lot of code changes recently so this needs a rebase plus addressing the rest of the feedback that you've gotten so far.

aw1875 commented:
> @aw1875 is this something that you plan on updating? There has been a lot of code changes recently so this needs a rebase plus addressing the rest of the feedback that you've gotten so far.

I've been so busy I haven't had time to take a look at the codebase in a while. I'll try to see if I can put some time aside tomorrow to re-familiarize myself with everything and look into fixing this up/updating things. If I can't find the time I'll probably close this pr out and come back to it when I do have more time if nobody has done it

aw1875 committed:
apprt/gtk: handle syncing background opacity at the opengl level

aw1875 commented:
Rebased and updated the code to fix the issue with not actually notifying OpenGL. It might make more sense to move the check for the value of `background-opacity` in the config to the App level so we can exit early rather running through the whole sync process when the value is already one. If I were to make this change, this is what I had in mind:

```diff
diff --git a/src/apprt/gtk/App.zig b/src/apprt/gtk/App.zig
index 22c3064f..a7a5bdfe 100644
--- a/src/apprt/gtk/App.zig
+++ b/src/apprt/gtk/App.zig
@@ -772,6 +772,8 @@ fn toggleBackgroundOpacity(
     switch (target) {
         .app => {},
         .surface => |v| {
+            if (self.config.@"background-opacity" >= 1) return;
+
             const window = v.rt_surface.container.window() orelse {
                 log.info(
                     "toggleBackgroundOpacity invalid for container={s}",

diff --git a/src/apprt/gtk/Window.zig b/src/apprt/gtk/Window.zig
index 9b1ba4c3..6ad094de 100644
--- a/src/apprt/gtk/Window.zig
+++ b/src/apprt/gtk/Window.zig
@@ -643,8 +643,6 @@ pub fn toggleWindowDecorations(self: *Window) void {

 /// Toggle the background opacity for this window.
 pub fn toggleBackgroundOpacity(self: *Window) !f64 {
-    if (self.app.config.@"background-opacity" >= 1) return self.app.config.@"background-opacity";
-
     self.config.background_opacity = if (self.config.background_opacity == 1.0) self.app.config.@"background-opacity" else 1.0;
     try self.syncAppearance();
```

aw1875 commented on src/apprt/gtk/Window.zig:
```diff
@@ -457,12 +457,6 @@ pub fn syncAppearance(self: *Window) !void {
     self.winproto.syncAppearance() catch |err| {
         log.warn("failed to sync winproto appearance error={}", .{err});
     };
-
-    toggleCssClass(
-        @ptrCast(self.window),
-        "background",
-        self.config.background_opacity >= 1,
-    );
```
Just as a note to reviewers, this was a duplicate call to the same toggle we do on line 422

pluiedev commented on src/apprt/gtk/App.zig:
```diff
@@ -796,6 +797,31 @@ fn toggleQuickTerminal(self: *App) !bool {
+            const opacity = try window.toggleBackgroundOpacity();
+
+            // Cycle through all surfaces and set the background opacity so they stay in sync
+            for (self.core_app.surfaces.items) |surface| {
+                surface.core_surface.setBackgroundOpacity(opacity);
+            }
```
Wait what? That's not right. It should only affect the core surface of the current apprt surface

aw1875 replied:
Wouldn't we want these changes reflected across the app? If I have several tabs open and I toggle the opacity wouldn't the expected behavior be to toggle it for them all?

pluiedev replied:
Then this should be an app-level action instead of a surface-level one

aw1875 replied:
would we still want to loop through the surfaces then? Or are you suggesting avoiding that?

jcollie replied:
Do we, in fact, want `toggle_background_opacity` to be an app-level toggle, rather than a per surface? If we had a per surface toggle, you can always make it an "app" toggle by prefixing the keybind will `all:` but there's no way to do the opposite.

jcollie replied:
Or, must this affect all surfaces in a window?

pluiedev replied:
I mean it makes the most sense for me for this to be a "window-level" action, but since we've rejected (I think?) that for core that means we have to stick with either app or surface level for now.

jcollie replied:
yeah i think it makes most sense to have the toggle affect only one surface. if users want different they can add `all:` to their keybind.

aw1875 replied:
Okay, so in the case of the `all` flag, we'd iterate through the surfaces as we do now; otherwise, we'd change the current surface?

jcollie replied:
Having the `all:` flag should mean the the Ghostty core will iterate though all of the surfaces for you, we only need to handle the targeted surface.

aw1875 replied:
One caveat I noticed here: since the changes are only handled at a surface level, it's pretty easy to get a mismatch of tabs. For example, if I have 2 tabs open and I use the toggle with the `all` flag, it will toggle those two open tabs to 100 opacity. Then, if I open a new tab, it will be at my default opacity and will work opposite to the two other tabs since the state is part of the derived config at the surface level.

pluiedev commented on src/apprt/gtk/Window.zig:
```diff
@@ -661,6 +661,16 @@ pub fn toggleWindowDecorations(self: *Window) void {
 
+/// Toggle the background opacity for this window.
+pub fn toggleBackgroundOpacity(self: *Window) !f64 {
+    if (self.app.config.@"background-opacity" >= 1) return self.app.config.@"background-opacity";
```
What's this for?

pluiedev replied:
Well if it is global state then it shouldn't be stored per-window. Should be part of the App's DerivedConfig instead.

aw1875 replied:
okay cool that sounds good. I wasn't sure if I should be modifying that for this logic but that makes sense, I can update that

aw1875 replied:
I might just be blind but I don't see a Derived Config at the app level, only the package config that you can set from modifying your ghostty config

pluiedev commented on src/apprt/gtk/Window.zig:
```diff
@@ -661,6 +661,16 @@ pub fn toggleWindowDecorations(self: *Window) void {
+pub fn toggleBackgroundOpacity(self: *Window) !f64 {
+    if (self.app.config.@"background-opacity" >= 1) return self.app.config.@"background-opacity";
+
+    self.config.background_opacity = if (self.config.background_opacity == 1.0) self.app.config.@"background-opacity" else 1.0;
```
Doing == on a float isn't a good idea. Use the `a - b < epsilon` method.

pluiedev commented on src/renderer/OpenGL.zig:
```diff
@@ -653,6 +653,17 @@ pub fn setVisible(self: *OpenGL, visible: bool) void {
+    defer if (single_threaded_draw) self.draw_mutex.unlock();
+
+    // We don't want to go below 0 or above 1
+    self.config.background_opacity = @max(0, @min(1, opacity));
```
std.math.clamp

aw1875 committed:
chore(gtk): stub out simple app level derived config

aw1875 committed:
chore: remove window level toggle and utilize std.math.clamp in opengl

pluiedev commented on src/apprt/gtk/App.zig:
```diff
@@ -52,6 +52,9 @@ pub const Options = struct {};
 config: Config,
 
+/// Derived configuration values that are computed from the main. Temporarily prefixed with `_` per https://discord.com/channels/1005603569187160125/1349556080090288232/1349569138846203955
+_config: DerivedConfig,
```
I'm not sure about the _ prefix - it usually indicates internal data other code shouldn't use, but we do intend to switch everything over to use derived config in the future. `derived_config` perhaps?

aw1875 commented:
Been a little too busy to keep catching up on rebasing, plus still waiting to hear back on some implementation details in [this discussion](https://github.com/ghostty-org/ghostty/discussions/6686), so going to close this pr out for the time being.
